### PR TITLE
Laboratory work #4, Elena Chevtaeva - 23FPL2

### PR DESCRIPTION
**Pull request by Helenhelen2005**

### CODE REVIEW

**Comment by Vasilisa-Blyudova** [comment]
No progress!

**Comment by Vasilisa-Blyudova** [comment]
No progress!

**Comment by Vasilisa-Blyudova** [comment]
no changes since last review!

File: lab_4_retrieval_w_clustering/main.py [hunk 1]
```diff
@@ -3,9 +3,17 @@
 
 Vector search with clusterization
 """
-
-# pylint: disable=undefined-variable, too-few-public-methods, unused-argument, duplicate-code, unused-private-member, super-init-not-called
-from lab_3_ann_retriever.main import BasicSearchEngine, Tokenizer, Vector, Vectorizer
+from lab_2_retrieval_w_bm25.main import calculate_bm25
+
+# pylint: disable=undefined-variable, too-few-public-methods, unused-argument,
```
Comment (Vasilisa-Blyudova) [inline lab_4_retrieval_w_clustering/main.py]: ignores before imports

File: lab_4_retrieval_w_clustering/main.py [hunk 2]
```diff
@@ -51,6 +65,13 @@ def set_tokenized_corpus(self, tokenized_corpus: TokenizedCorpus) -> None:
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if (not isinstance(self._corpus, list) or tokenized_corpus == []
+                or not all(isinstance(doc, list) for doc in self._corpus)):
+            raise ValueError
```
Comment (Vasilisa-Blyudova) [inline lab_4_retrieval_w_clustering/main.py]: add message everywhere

File: lab_4_retrieval_w_clustering/main.py [hunk 3]
```diff
@@ -81,6 +109,18 @@ def _calculate_bm25(self, tokenized_document: list[str]) -> Vector:
         Returns:
             Vector: BM25 vector for document.
         """
+        if (not isinstance(tokenized_document, list) or not tokenized_document or
+                not all(isinstance(token, str) for token in tokenized_document)):
+            raise ValueError('Inappropriate type input argument / argument')
+
+        vector = list(0.0 for _ in range(len(self._vocabulary)))
```
Comment (Vasilisa-Blyudova) [inline lab_4_retrieval_w_clustering/main.py]: try to avoid loop here

File: lab_4_retrieval_w_clustering/main.py [hunk 4]
```diff
@@ -81,6 +109,18 @@ def _calculate_bm25(self, tokenized_document: list[str]) -> Vector:
...
+        bm25_values = calculate_bm25(self._vocabulary, tokenized_document, self._idf_values,
+                                     1.5, 0.75, self._avg_doc_len, len(tokenized_document))
```
Comment (Vasilisa-Blyudova) [inline lab_4_retrieval_w_clustering/main.py]: rm default arguments

File: lab_4_retrieval_w_clustering/main.py [hunk 5]
```diff
@@ -182,6 +261,23 @@ def retrieve_relevant_documents(self, query: str, n_neighbours: int) -> list[tup
         Returns:
             list[tuple[float, str]]: Relevant documents with their distances.
         """
+        if (not query or not isinstance(query, str)
+                or n_neighbours <= 0 or not isinstance(n_neighbours, int)):
```
Comment (Vasilisa-Blyudova) [inline lab_4_retrieval_w_clustering/main.py]: can it really be 0?

File: lab_4_retrieval_w_clustering/main.py [hunk 6]
```diff
@@ -271,11 +381,20 @@ def __init__(self, db: DocumentVectorDB, n_clusters: int) -> None:
...
+        self._n_clusters = n_clusters
 
     def train(self) -> None:
         """
         Train k-means algorithm.
         """
+        new_centroids = self._db.get_vectors()[:self._n_clusters]
+        self.__clusters = [ClusterDTO(centroid[1]) for centroid in new_centroids]
+        self.run_single_train_iteration()
+        while self._is_convergence_reached(self.__clusters) is False:
```
Comment (Vasilisa-Blyudova) [inline lab_4_retrieval_w_clustering/main.py]: don't use `is False`

File: lab_4_retrieval_w_clustering/main.py [hunk 7]
```diff
@@ -271,11 +381,20 @@ def __init__(self, db: DocumentVectorDB, n_clusters: int) -> None:
...
+            self.run_single_train_iteration()
+            self._is_convergence_reached(self.__clusters)
```
Comment (Vasilisa-Blyudova) [inline lab_4_retrieval_w_clustering/main.py:407]: is it really need?

File: lab_4_retrieval_w_clustering/main.py [hunk 8]
```diff
@@ -343,6 +512,17 @@ def _is_convergence_reached(
         Returns:
             bool: True if the distance is correct, False in other cases.
         """
+        if not new_clusters:
+            raise ValueError('New clusters must be non-empty. ')
+
+        for ind, old_cluster in enumerate(self.__clusters):
```
Comment (Vasilisa-Blyudova) [inline lab_4_retrieval_w_clustering/main.py]: use `zip()` in loop

File: lab_4_retrieval_w_clustering/main.py [hunk 9]
```diff
@@ -182,6 +261,23 @@ def retrieve_relevant_documents(self, query: str, n_neighbours: int) -> list[tup
         Returns:
             list[tuple[float, str]]: Relevant documents with their distances.
         """
+        if (not query or not isinstance(query, str)
+                or n_neighbours <= 0 or not isinstance(n_neighbours, int)):
+            raise ValueError('Query must be a non-empty string; '
+                             'Number of neighbours must be a positive integer.')
+
+        tokenized_query = self._db.get_tokenizer().tokenize(query)
```
Comment (Vasilisa-Blyudova) [inline lab_4_retrieval_w_clustering/main.py:278]: use tokenizer and vectorizer from parent class

**Review by Vasilisa-Blyudova — COMMENTED** [review]
Good work!

File: lab_4_retrieval_w_clustering/main.py (hunk 1 above)
Comment (Helenhelen2005) [reply lab_4_retrieval_w_clustering/main.py]: fixed

File: lab_4_retrieval_w_clustering/main.py (hunk 3 above)
Comment (Helenhelen2005) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 4 above)
Comment (Helenhelen2005) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 6 above)
Comment (Helenhelen2005) [reply lab_4_retrieval_w_clustering/main.py]: done (used not True instead, not sure it is what you meant...)

File: lab_4_retrieval_w_clustering/main.py (hunk 7 above)
Comment (Helenhelen2005) [reply lab_4_retrieval_w_clustering/main.py:407]: i'm sorry, i don't get the comment...

File: lab_4_retrieval_w_clustering/main.py (hunk 8 above)
Comment (Helenhelen2005) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 5 above)
Comment (Helenhelen2005) [reply lab_4_retrieval_w_clustering/main.py]: fixed

File: lab_4_retrieval_w_clustering/main.py (hunk 2 above)
Comment (Helenhelen2005) [reply lab_4_retrieval_w_clustering/main.py]: in progress..

Comment (Helenhelen2005) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 7 above)
Comment (Helenhelen2005) [reply lab_4_retrieval_w_clustering/main.py:407]: now i got it
done

**Pull request closed without merging**